client/chat/channel-settings-dialog.tsx

```diff
@@ -45,106 +76,178 @@ export function ChannelSettingsDialog({
+
+  const bannerInputRef = useRef<FileInputHandle>(null)
+  const badgeInputRef = useRef<FileInputHandle>(null)
+
+  const onFormSubmit = useStableCallback((model: Readonly<ChannelSettingsModel>) => {
+    const patch: EditChannelRequest = {
+      description:
+        model.description !== detailedChannelInfo.description ? model.description : undefined,
+      topic: model.topic !== joinedChannelInfo.topic ? model.topic : undefined,
+      deleteBanner: detailedChannelInfo.bannerPath && !model.banner ? true : undefined,
+      deleteBadge: detailedChannelInfo.badgePath && !model.badge ? true : undefined,
+    }
+
+    dispatch(
+      updateChannel({
+        channelId: basicChannelInfo.id,
+        channelChanges: patch,
+        channelBanner: model.banner instanceof File ? model.banner : undefined,
+        channelBadge: model.badge instanceof File ? model.badge : undefined,
+        spec: {
+          onSuccess: () => {},
+          onError: err => {
+            dispatch(
+              openSnackbar({
+                message: t(
+                  'chat.channelSettings.general.saveErrorMessage',
+                  'Something went wrong saving the settings',
+                ),
+              }),
+            )
+          },
+        },
+      }),
+    )
   })
-  const [buttonProps, rippleRef] = useButtonState({ onClick })
 
-  let title
-  switch (section) {
-    case ChannelSettingsSection.General:
-      title = t('chat.channelSettings.general.title', 'General')
-      break
-    default:
-      assertUnreachable(section)
-  }
+  const { bindCustom, bindInput, getInputValue, setInputValue, onSubmit } = useForm(
+    {
+      description: detailedChannelInfo.description,
+      topic: joinedChannelInfo.topic,
+      banner: detailedChannelInfo.bannerPath,
+      badge: detailedChannelInfo.badgePath,
+    },
+    {},
+    { onSubmit: onFormSubmit },
+  )
 
-  return (
-    <NavEntryRoot $isActive={isActive} {...buttonProps} tabIndex={0}>
-      <NavEntryText>{title}</NavEntryText>
+  const bannerUrl = useObjectUrl(getInputValue('banner'))
+  const badgeUrl = useObjectUrl(getInputValue('badge'))
+
+  const buttons = [
+    <TextButton
+      label={t('common.actions.cancel', 'Cancel')}
+      key='cancel'
+      color='accent'
+      onClick={() => dispatch(closeDialog(DialogType.ChannelSettings))}
+    />,
+    <TextButton
+      label={t('common.actions.save', 'Save')}
+      key='save'
+      color='accent'
+      onClick={() => {
+        onSubmit()
+        dispatch(closeDialog(DialogType.ChannelSettings))
+      }}
+    />,
+  ]
 
-      <Ripple ref={rippleRef} />
-    </NavEntryRoot>
+  return (
+    <StyledDialog dialogRef={dialogRef} title={`#${basicChannelInfo?.name}`} buttons={buttons}>
+      <Root>
+        <FormContainer>
+          <StyledForm noValidate={true} onSubmit={onSubmit}>
+            <BannerButtonsContainer>
+              <HiddenFileInput
```

Comment:
I don't think this is really how this should be handled. The file input component should provide a button or otherwise, but still be visible (e.g. definitely *not* `display: none`) so that things like file dragging work on it. If you deal with things properly then implementing those behaviors is free